controllers/lighting_controller/lighting_controller.py

```diff
@@ -0,0 +1,270 @@
+
+            print(f"Lighting effect '{effect.name}' complete")  # noqa: T001
+
+        else:
+            steps = int((effect.fade_time * 1000) / self.timestep)
+
+            # get starting values
+            current_luminosity = self.get_current_luminosity()
+            luminosity_step = (effect.luminosity - current_luminosity) / steps
+            self.luminosity_fade = LuminosityFade(
+                steps,
+                luminosity_step,
+                current_luminosity,
+                effect.luminosity,
+            )
+
+            for light in effect.lighting:
+                # get starting values
+                (
+                    _,
+                    current_intensity,
+                    current_colour,
+                ) = self.get_current_lighting_values(light.light_def)
+
+                # figure out steps of each value
+                intensity_step = (light.intensity - current_intensity) / steps
+                colour_step: Tuple[float, float, float] = tuple(  # type: ignore
+                    light.colour[i] - current_colour[i]
+                    for i in range(3)
+                )
+
+                # add fade to queue to have steps processed
+                self.lighting_fades.append(LightFade(
+                    self.light_nodes[light.light_def],
+                    steps,
+                    intensity_step,
+                    colour_step,
+                    current_intensity,
+                    current_colour,
+                    light,
+                ))
+
+    def do_lighting_step(self) -> None:
+        if self.luminosity_fade.remaining_steps != 0:
+            self.luminosity_fade.current_luminosity += self.luminosity_fade.luminosity_step
+            self.luminosity_fade.remaining_steps -= 1
+
+            if self.luminosity_fade.remaining_steps == 0:  # final step
+                self.luminosity_fade.current_luminosity = self.luminosity_fade.final_luminosity
+
+            self.set_node_luminosity(
+                self.ambient_node,
+                self.luminosity_fade.current_luminosity,
+            )
+
+        for fade in self.lighting_fades:
+            if fade.remaining_steps > 1:
+                fade.current_intensity += fade.intensity_step
+                fade.current_colour = self.increment_colour(
+                    fade.current_colour,
+                    fade.colour_step,
+                )
+                fade.remaining_steps -= 1
+
+                self.set_node_intensity(fade.light, fade.current_intensity)
+                self.set_node_colour(fade.light, fade.current_colour)
+            else:
+                self.set_node_intensity(fade.light, fade.effect.intensity)
+                self.set_node_colour(fade.light, fade.effect.colour)
+
+                print(f"Lighting effect for '{fade.effect.light_def}' complete")  # noqa: T001
+
+                self.lighting_fades.remove(fade)  # remove completed fade
```

Comment:
Does this not change the size during iteration, and thus cause issues?

Reply:
Indeed, fixed in f428b561f20fcf9c2c40ffa07c463d4a820a0b74